### PR TITLE
[c++] Implementation and unit testing for domainish accessors

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** As tracked on issue #2407 / [[sc-51048]](https://app.shortcut.com/tiledb-inc/story/51048).

This is atop #3011; these will replace #2995.

Note that the intended Python and R API changes are all agreed on and finalized as described in #2407.

**Changes:**

**Notes for Reviewer:**

Python/R PRs will be separate.

### CODE REVIEW

File: libtiledbsoma/src/soma/soma_array.h [hunk 1]
```diff
@@ -775,6 +847,108 @@ class SOMAArray : public SOMAObject {
...
+        switch (which_kind) {
+            case Domainish::kind_core_domain:
+                return _core_domain_slot<T>(name);
+                break;
+            case Domainish::kind_core_current_domain:
+                return _core_current_domain_slot<T>(name);
+                break;
+            case Domainish::kind_non_empty_domain:
+                return non_empty_domain_slot<T>(name);
+                break;
```
Comment (nguyenv) [inline libtiledbsoma/src/soma/soma_array.h:927]:
```suggestion
            case Domainish::kind_core_domain:
                return _core_domain_slot<T>(name);
            case Domainish::kind_core_current_domain:
                return _core_current_domain_slot<T>(name);
            case Domainish::kind_non_empty_domain:
                return non_empty_domain_slot<T>(name);
```

File: libtiledbsoma/src/soma/soma_array.h [hunk 2]
```diff
@@ -775,6 +847,108 @@ class SOMAArray : public SOMAObject {
...
+            case Domainish::kind_core_current_domain:
+                return _core_current_domain_slot<std::string>(name);
+                break;
+            case Domainish::kind_non_empty_domain:
+                return non_empty_domain_slot_var(name);
+                break;
+            default:
+                throw std::runtime_error(
+                    "internal coding error in SOMAArray::_core_domainish_slot");
```
Comment (nguyenv) [inline libtiledbsoma/src/soma/soma_array.h:948]: Do we also want to elaborate that the internal coding error is an unknown `Domainish` type?

```suggestion
                    "internal coding error in SOMAArray::_core_domainish_slot_string");
```

File: libtiledbsoma/src/soma/soma_array.h [hunk 3]
```diff
@@ -775,6 +847,108 @@ class SOMAArray : public SOMAObject {
...
+        switch (which_kind) {
+            case Domainish::kind_core_domain:
+                return _core_domain_slot_string(name);
+                break;
+            case Domainish::kind_core_current_domain:
+                return _core_current_domain_slot<std::string>(name);
+                break;
+            case Domainish::kind_non_empty_domain:
+                return non_empty_domain_slot_var(name);
+                break;
```
Comment (nguyenv) [inline libtiledbsoma/src/soma/soma_array.h:945]:
```suggestion
            case Domainish::kind_core_domain:
                return _core_domain_slot_string(name);
            case Domainish::kind_core_current_domain:
                return _core_current_domain_slot<std::string>(name);
            case Domainish::kind_non_empty_domain:
                return non_empty_domain_slot_var(name);
```

File: libtiledbsoma/src/soma/soma_array.cc [hunk 4]
```diff
@@ -1141,6 +1140,105 @@ std::optional<TimestampRange> SOMAArray::timestamp() {
...
+            default:
+                break;
+        }
+
+        if (child == nullptr) {
+            // throw TileDBSOMAError(fmt::format(
+            //   "WIP {} {}",
+            //   core_dim.name(),
+            //   tiledb::impl::type_to_str(core_type_code)));
+        }
```
Comment (nguyenv) [inline libtiledbsoma/src/soma/soma_array.cc:1235]: Did you intend to throw here if the `core_type_code` was not found or set the child to `nullptr`?

File: libtiledbsoma/src/soma/soma_array.h (hunk 2 above)
Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_array.h:948]: good call! :)

File: libtiledbsoma/src/soma/soma_array.cc (hunk 4 above)
Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_array.cc:1235]: derp sorry, that's a WIP I didn't finish & I shouldn't have put this PR up for review yet

Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_array.cc]: https://github.com/single-cell-data/TileDB-SOMA/pull/3012/commits/804db2b41d3c4f31e41fd08bbc55b5730c79a794

File: libtiledbsoma/src/soma/soma_array.cc [hunk 5]
```diff
@@ -1141,6 +1140,105 @@ std::optional<TimestampRange> SOMAArray::timestamp() {
...
+
+            default:
+                break;
+        }
+
+        if (child == nullptr) {
+            throw TileDBSOMAError(fmt::format(
+                "SOMAArray::_get_core_domainish:dim {} has unhandled type {}",
+                core_dim.name(),
+                tiledb::impl::type_to_str(core_type_code)));
```
Comment (nguyenv) [inline libtiledbsoma/src/soma/soma_array.cc:1234]:
```suggestion
            default:
                throw TileDBSOMAError(fmt::format(
                    "SOMAArray::_get_core_domainish:dim {} has unhandled type {}",
                     core_dim.name(),
                     tiledb::impl::type_to_str(core_type_code)));
```

**Review by nguyenv — APPROVED** [review]

**Pull request merged**